### ISSUE TITLE
Question about I2S_CLK and BCK generation

### ISSUE THREAD

**Issue by danjulio** [aquaticus/esp32_composite_video_lib#11]
As part of a port of the library to IDF 5.3 I have been doing a deep dive into your code and I have a couple of questions about how the I2S clocks are generated (the values you are loading into the I2S_CLKM_CONF_REG and I2S_SAMPLE_RATE_CONF_REG).

Take normal NTSC video for example (6.136 MHz clock rate).  It looks to me like you configure the APLL to output 24.5454 MHz (4X) which is then fed into the I2S Clock generator.  I think you want to have the W clock rate to be the original 6.136 MHz and the ESP32 ref manual says that W is half of BCK so I think you want BCK to be 12.2727 MHz ("Camera-LCD Controller" section 12.5 page 315).  This would imply that between the I2S_CLK generator (Fapll / (N + b/a) and the BCK geneator (I2S_CLK / M) you want to divide by 2.

However I see that you set N = 1, b = 0, a = 1, M = 1 in the code

```
    I2S0.clkm_conf.clkm_div_num = 1;            // I2S clock divider's integral value.
    I2S0.clkm_conf.clkm_div_b = 0;              // Fractional clock divider's numerator value.
    I2S0.clkm_conf.clkm_div_a = 1;              // Fractional clock divider's denominator value
    I2S0.sample_rate_conf.tx_bck_div_num = 1;
```

This implies that both I2S_CLK and BCK would be 24.5454 MHz (both divisions are by 1).  Furthermore the spec ambiguously says both N and M values are >= 2 ("The Clock of the I2S Module" section 12.3 page 307).  So I am confused.  I don't understand why you are setting them to 1 and I don't understand how you get the correct W clock to actually feed the pixels at the NTSC frequency of 6.136 MHz.

Would you mind explaining what's going on here and/or if I have made some errors in my analysis?

My references to the ESP32 reference manual are for version 4.8.

**Comment by aquaticus** [comment]
Hi,

The clock frequency is actually set in function https://github.com/aquaticus/esp32_composite_video_lib/blob/a1f5c7669aa274f148157947cdb7c94459e23777/video.c#L253

To set frequency you call `rtc_clk_apll_enable()` function. It configures Audio PLL clock which is derived from system clock ($f_{xtal}=40MHz$). You must feed `rtc_clk_apll_enable()` with the divider which gives you the closest frequency to video signal clock.

The output frequency of Audio PLL can be calculated using the following formula:


$`f_{out}=\frac{f_{xtal}(sdm2+\frac{sdm1}{2^{8}}+\frac{sdm0}{2^{16}}+4)}{2(odiv+2)}`$

See  [ESP32 Technical Reference](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf) page 45 - 3.2.7 Audio PLL for more details.

**Comment by danjulio** [comment]
Thanks, I understand what you wrote.  I understand the APLL setup.  What I don't understand is the configuration of the I2S (not the APLL).  Specifically I don't know why you set I2S_CLKM_CONF_REG and I2S_SAMPLE_RATE_CONF_REG the way you do in `setup_video_dac()`.  I see that [for NTSC] you are generating a 24.5454 MHz APLL clock but I don't understand how you end up with a 6.136 MHz dot clock through the I2S divisor chain.

**Comment by aquaticus** [comment]
Please note the line https://github.com/aquaticus/esp32_composite_video_lib/blob/a1f5c7669aa274f148157947cdb7c94459e23777/video.c#L320

It configures I2S to use Audio PLL clock. And that's all. The rest of the I2S freq config does not matter in that case (as far as I remember :-) ).  See page 315 _12.3 The Clock of I2S Module_.

The next step is to configure Audio PLL but you already know that.

**Comment by aquaticus** [comment]
Sorry, one update.
```
  I2S0.clkm_conf.clkm_div_num = 1;            // I2S clock divider's integral value.
  I2S0.clkm_conf.clkm_div_b = 0;              // Fractional clock divider's numerator value.
  I2S0.clkm_conf.clkm_div_a = 1;              // Fractional clock divider's denominator value
```

it actually matters. It configures divider as shown on the diagram on page 315.
But resulting divider is 1, so APP_CLK is not modified

**Comment by danjulio** [comment]
I really appreciate you taking the time to look at this and I feel stupid because something is not making sense for me.  If you can bear with me for another round.

I compute your APLL clock to be 24.5454 MHz based on the constants in your code.  If the I2S divisors are 1 then both the I2S_CLK and the BCK are also going to be 24.5454 MHz.  Figure 12-19 (page 317) shows the DAC being loaded on WS clock intervals (even if I2S_CLK is faster).  And the spec (page 315) says WS is half of BCK so this implies WS would be 12.2727 MHz which is twice as fast as it should be (6.136 MHz) to clock the dots out.  

In other words: I don't get how you end up clocking data out to the DAC at 6.136 MHz given your setup.  It seems to me it should be coming out twice as quickly.

And curious is that the spec says both the clkm_div_num and tx_bck_div_num should be >= 2 (page 307).  Maybe this doesn't actually matter to the chip but I still don't get how the divide by 1 works.


> ”N”, whose value is >=2, corresponds to the REG _CLKM_DIV_NUM [7:0] bits of register I2S_CLKM_CONF_REG
> 
> ”M”, whose value is >=2, is the I2S_TX_BCK_DIV_NUM[5:0] bit of register I2S_SAMPLE _RATE_CONF_REG

**Comment by aquaticus** [comment]
ok, now I got what you mean. Sorry for my ignorance.
Unfortunately, I don't remember the answer.

I remember I made note on that on Python script I used to generate values for `rtc_clk_apll_enable()` (it seems to be lost). I guess that somehow `N` value is equal to 4. It could be a bug in documentation. I remember there was yet another document discussing only audio aspects of ESP32 I was using.

Sorry but at the moment I cannot give you any valuable explanation. Maybe I'll check my archives tomorrow but no guarantee.

**Comment by danjulio** [comment]
Ok.  No worries if you can't find it.  My experiment trying to use the native IDF 5.3 DAC driver seems to have failed so I'm going back to your code base.  I will play with those settings once I have it running to try to see if I can learn anything.

Thanks again, for replying, and for creating this wonderful [mis]use of the ESP32 DAC :-)

**Comment by aquaticus** [comment]
I checked my backups but could not find anything valuable.
If you ever find why this work like that, please let me know :-)

**Comment by danjulio** [comment]
Thanks for looking.  I played with the settings and as you'd expect changing either of the I2S clock generator slowed the output down by the same divisor (and if I increased APLL clock by the same ratio it worked again).  So I must not understand how the APLL clock feeds the DAC I2S output.  However your values work so I'm done worrying about this and will close this.  I'll update it if I ever figure something out.
